module/zfs/zfs_vnops.c

```diff
@@ -1292,10 +1292,18 @@ zfs_clone_range(znode_t *inzp, uint64_t *inoffp, znode_t *outzp,
 		if (error != 0) {
 			/*
 			 * If we are trying to clone a block that was created
-			 * in the current transaction group, error will be
-			 * EAGAIN here, which we can just return to the caller
-			 * so it can fallback if it likes.
+			 * in the current transaction group, the error will be
+			 * EAGAIN here.  Based on sync_wait either return a
+			 * shortened range to the caller so it can fallback,
+			 * or wait for the next sync and check again.
 			 */
+			if (error == EAGAIN && wait_sync) {
+				int64_t txg = spa_last_synced_txg(
```

Comment:
As a small optimization - maybe fetch `txg` before `dmu_read_l0_bps`?

In the case where a sync happens to complete in the background _during_ the call to `dmu_read_l0_bps`, then this will retry immediately instead of triggering a sync. If there are still dirty blocks then the following iteration will also return `EAGAIN` and sync again here.

Reply:
Done.